rclcpp/include/rclcpp/publisher_options.hpp

```diff
@@ -102,7 +103,7 @@ struct PublisherOptionsWithAllocator : public PublisherOptionsBase
   get_allocator() const
   {
     if (!this->allocator) {
-      return std::make_shared<Allocator>();
+      this->allocator = std::make_shared<Allocator>();
```

Comment:
I'm not sure this is such a good idea. This would be a weird side-effect of `to_rcl_publisher_options`.

Instead, maybe should leave this as-is (since this methood's purpose is to provide an allocator, even if one is not set, but not to ensure it stays alive). And then we can have logic in `to_rcl_publisher_options` like this:

```c++
if (this->allocator) {
  result.allocator = rclcpp::allocator::get_rcl_allocator<void>(*this->get_allocator());
} else {
  result.allocator = rcl_get_default_allocator();
}
```

Or perhaps:

```c++
if (this->allocator) {
  result.allocator = rclcpp::allocator::get_rcl_allocator<void>(*this->get_allocator());
} else {
  static std::allocator<void> default_std_allocator;
  result.allocator = rclcpp::allocator::get_rcl_allocator<void>(default_std_allocator);
}
```

Reply:
unrelated question, does `std::shared_ptr<Allocator>` makes sense at all?

IIUC allocators copy constructors must not throw and must "maintain the same resources" as the original one.
This also apply when constructing from a rebinded version.

From [cppreference](https://en.cppreference.com/w/cpp/named_req/Allocator):

Declaration | Effect | Requirements
-- | -- | --
A a1(a); <br> A a1 = a; | Copy-constructs a1 such that a1 == a. (Note: Every Allocator also satisfies CopyConstructible.) | Does not throw exceptions.
A a(b) | Constructs a such that B(a)==b and A(b)==a. (Note: This implies that all allocators related by rebind maintain each other's resources, such as memory pools.) | Does not throw exceptions.

i.e.: If the allocator implementation is stateful, it seems to already be a requirement that the internal state may be shared by difference allocator instances.

Reply:
> ```c++
> if (this->allocator) {
>  result.allocator = rclcpp::allocator::get_rcl_allocator<void>(*this->get_allocator());
>} else {
>  result.allocator = rcl_get_default_allocator();
>}
>```

The problem with this options is that IDK what to do in places like here:

https://github.com/ros2/rclcpp/blob/79c2dd8e8b3e73366c59474b91271a2fd57954bc/rclcpp/include/rclcpp/subscription_factory.hpp#L93

if `allocator` is `nullptr`, it doesn't sound like a good idea to create a new instance because that might involve creating new resources in custom allocators (though that sounds valid for `std::allocator`).

> Or perhaps:
> 
> ```c++
> if (this->allocator) {
>   result.allocator = rclcpp::allocator::get_rcl_allocator<void>(*this->get_allocator());
>} else {
>  static std::allocator<void> default_std_allocator;
>  result.allocator = rclcpp::allocator::get_rcl_allocator<void>(default_std_allocator);
>}
>```

That one sounds valid for a `std::allocator`, but it doesn't for custom ones (because the static allocator might be using resources unnecessarily).

Maybe I can specialize the method for `std::allocator<void>`, which returns an static instance of it.
For custom allocators, an exception is thrown if `nullptr`.

Reply:
See https://github.com/ros2/rclcpp/pull/1647/commits/6ff1246a3de528ee829628453adac269a2b25709.

(addressing comments here because Mich is OoO and wanted to get this one in)

Reply:
> unrelated question, does std::shared_ptr<Allocator> makes sense at all?

No, see: https://github.com/ros2/rclcpp/issues/849#issuecomment-539758462

I wanted to remove all use of `shared_ptr<AllocatorT>`, but haven't been able to justify the churn caused by it. We should do it though, and I have been doing it as I go, where I can.

> if allocator is nullptr, it doesn't sound like a good idea to create a new instance because that might involve creating new resources in custom allocators (though that sounds valid for std::allocator).

I don't think that's an issue. If the allocator is nullptr, but the template argument is a custom allocator, then the caller has to know we'll create one for them. It's no different than the std::vector constructor, for example:

https://en.cppreference.com/w/cpp/container/vector/vector

Notice that the ones that take the allocator as an argument have a default value of `const Allocator& alloc = Allocator()`, so us creating one when it is needed and the member is nullptr, really isn't different.

Long term, maybe having a constructor for the options that optionally takes an allocator is the right thing to do, e.g.:

```c++
PublisherOptionsWithAllocator(const Allocator & allocator = Allocator())
: allocator(std::make_shared<Allocator>(allocator)) {}
```

Later we'd want to change the member of the options to be a non-`shared_ptr`.

Reply:
> I don't think that's an issue. If the allocator is nullptr, but the template argument is a custom allocator, then the caller has to know we'll create one for them. It's no different than the std::vector constructor, for example:

The difference is that we call `*Options.get_allocator()` multiple times, not only once.
So when `options.allocator = nullptr`, we're going to create multiple new instances of the allocator type instead of only one and copies of that one when `options.allocator != nullptr`.

The version using a `static` works if the compiler would guarantee that they are initialized lazely, but that's not guaranteed and if `options.allocator != nullptr` we may create two instances of the allocator (which might matter if the allocator is stateful).

Reply:
> The difference is that we call *Options.get_allocator() multiple times, not only once.

Actually, I think we might be calling these methods only once.
I will double check on Monday ...

Reply:
If we're calling it only once, I'd avoid throwing. If we're calling more than once I'd advocate for a second storage for the "default", construct it lazily, then return it from then on.

Either way, I think default constructing one when the user doesn't provide one is ok, it's just the current architecture leaves us in a weird place, as I said above about the constructor.

Reply:
> second storage

Does adding a new private member variable for that sound ok? Seems weird, but I'm ok with it.
Besides that, the only other options I see is adding a `mutable` keyword to the `allocator` member as @hidmic tried before and a `static`.

> Either way, I think default constructing one when the user doesn't provide one is ok, it's just the current architecture leaves us in a weird place, as I said above about the constructor.

---

Defualt constructing one seems okay to me.
The only problem is that I guess that most code using allocators is currently doing:

```cpp
PublisherOptionsWithAllocator options;
options.allocator = <expression>;
```

and that would now default construct an allocator + destroy that one and set `options.allocator` to <expression>, which may be undesirable.

Long term, we might want to add:

```cpp
  PublisherOptions(const & Allocator alloc = Allocator());
```

and make the `allocator` member variable private (`get_allocator()` will be the getter).
It will break all code that uses allocators now, but that's not too much code and I think it's better to break code hard if we want to change the way allocators are initialized in `PublisherOptions`/`SubscriptionOptions`.
We could probably take advantage of the breakage there and also change `std::shared_ptr<Allocator>` to `Allocator`.

Reply:
> Does adding a new private member variable for that sound ok? Seems weird, but I'm ok with it.

I don't think it's an issue, to be honest. Like I said, this is a workaround, long term we should use a constructor argument and get away from shared ptr too.

> Defualt constructing one seems okay to me.
The only problem is that I guess that most code using allocators is currently doing:

Yeah, I'd avoid that, and go for the private default, and have `get_allocator()` return one or the other.

> Long term, we might want to add:

Agreed.